Tests/TestUtils/TestNetworkService.swift

```diff
@@ -0,0 +1,190 @@
+//
+// Copyright 2020 Adobe. All rights reserved.
+// This file is licensed to you under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License. You may obtain a copy
+// of the License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under
+// the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+// OF ANY KIND, either express or implied. See the License for the specific language
+// governing permissions and limitations under the License.
+//
+
+@testable import AEPEdge
+@testable import AEPServices
+import Foundation
+import XCTest
+
+/// DO NOT use this class directly in tests. Use the child classes, either `MockTestNetworkService` or
+/// `ServerTestNetworkService` depending on which is appropriate for your use case.
+/// The base NetworkService class that implements shared utilities and logic for NetworkService class implementations
+/// used for testing.
+class TestNetworkService: NetworkService {
+    private var sentNetworkRequests: [NetworkRequest: [NetworkRequest]] = [:]
+    /// Matches sent `NetworkRequest`s with their corresponding `HttpConnection` response.
+    private(set) var networkResponses: [NetworkRequest: HttpConnection] = [:]
+    private var expectedNetworkRequests: [NetworkRequest: CountDownLatch] = [:]
+    
+    func recordSentNetworkRequest(_ networkRequest: NetworkRequest) {
+        TestBase.log("Received connectAsync to URL \(networkRequest.url.absoluteString) and HTTPMethod \(networkRequest.httpMethod.toString())")
+        if let equalNetworkRequest = sentNetworkRequests.first(where: { key, value in
+            areNetworkRequestsEqual(lhs: networkRequest, rhs: key)
+        }) {
+            sentNetworkRequests[equalNetworkRequest.key]!.append(networkRequest)
+        }
+        else {
+            sentNetworkRequests[networkRequest] = [networkRequest]
+        }
+    }
+
+    func reset() {
+        expectedNetworkRequests.removeAll()
+        sentNetworkRequests.removeAll()
+        networkResponses.removeAll()
+    }
+    
+    /// Equals compare based on host, scheme and URL path. Query params are not taken into consideration
+    func areNetworkRequestsEqual(lhs: NetworkRequest, rhs: NetworkRequest) -> Bool {
```

Comment:
nit: We can also extend the NetworkRequest class and inherit Comparable to its definition and implement == (equals) function for it?

Reply:
Create a file in your test target Network+Equatable

```
extension NetworkRequest: Equatable {
static func == (lhs: NetworkRequest, rhs: NetworkRequest) -> Bool {
       //code
    }
}
```

let me know if this works.

Reply:
we can at least add it to the NetworkRequest extension so usages can look like
func isEqual(_ other: NetworkRequest) -> Bool { .. }
usage: networkRequest.isEqual(otherNetworkRequest)